### PR TITLE
More tests for replace with AbstractDict

### PR DESCRIPTION
**Pull request by kshyatt**

### CODE REVIEW

**Comment by Seelengrab** [comment]
What do these tests test/ensure?

**Comment by kshyatt** [comment]
These functions are basically [totally untested](https://app.codecov.io/gh/JuliaLang/julia/blob/master/base%2Fset.jl#L739) for `AbstractDict` subtypes that aren't `Dict`.

**Comment by Seelengrab** [comment]
Increasing coverage is pretty much always good, yes!

File: test/sets.jl [hunk 1]
```diff
@@ -719,6 +719,18 @@ end
...
+    interior::Dict{K, V}
+end
+Base.setindex!(rd::ReplaceDict, v, k) = setindex!(rd.interior, v, k)
+Base.getindex(rd::ReplaceDict, k) = getindex(rd.interior, k)
+Base.get(rd::ReplaceDict, k, default) = get(rd.interior, k, default)
+Base.iterate(rd::ReplaceDict) = iterate(rd.interior)
+Base.iterate(rd::ReplaceDict, x) = iterate(rd.interior, x)
+Base.length(rd::ReplaceDict) = length(rd.interior)
+Base.get!(rd::ReplaceDict, k, default) = get!(rd.interior, k, default)
+Base.pop!(rd::ReplaceDict, x) = pop!(rd.interior, x)
```
Comment (rfourquet) [inline test/sets.jl:733]: Could you use `Test.GenericDict` instead?

File: test/sets.jl [hunk 2]
```diff
@@ -810,6 +822,19 @@ end
     @test replace((NaN, 1.0), NaN=>0.0) === (0.0, 1.0)
     @test replace([1, missing], missing=>0) == [1, 0]
     @test replace((1, missing), missing=>0) === (1, 0)
+
+    # test replace and friends for AbstractDicts
+    d1 = ReplaceDict(Dict(1=>2, 3=>4))
+    d2 = replace(d1, (1=>2) => (1=>"a"))
+    @test d2 == ReplaceDict(Dict(1=>"a", 3=>4))
```
Comment (rfourquet) [inline test/sets.jl:829]: this doesn't test that `d2` is a `ReplaceDict`:
```suggestion
    @test d2 == Dict(1=>"a", 3=>4)
    @test d2 isa ReplaceDict
```

File: test/sets.jl [hunk 3]
```diff
@@ -810,6 +822,19 @@ end
...
+    replace!(d1, (1=>2) => (1=>-2))
+    @test d1 == ReplaceDict(Dict(1=>-2, 3=>4))
+
+    d1 = ReplaceDict(Dict(1=>2, 3=>1, 5=>1, 7=>1))
+    @test replace(d1, (1=>2) => (1=>"a"), count=0) == d1
+    d2 = replace((kv)->(kv[2] == 1 ? kv[1]=>2 : kv), d1, count=2)
+    @test count(v->v==2, values(d2)) == 3
+    @test count(v->v==1, values(d2)) == 1
```
Comment (rfourquet) [inline test/sets.jl:837]: It's subjective comment (feel free to ignore), but it's hard to follow whether this property was already tested for `Dict`. If it was not, I would try to make this test for both types, e.g. : 
```suggestion
    dd = Dict(1=>2, 3=>1, 5=>1, 7=>1)
    for d1 = (dd, ReplaceDict(dd))
        @test replace(d1, (1=>2) => (1=>"a"), count=0) == d1
        d2 = replace((kv)->(kv[2] == 1 ? kv[1]=>2 : kv), d1, count=2)
        @test count(v->v==2, values(d2)) == 3
        @test count(v->v==1, values(d2)) == 1
    end
```
It it was already tested, i would similarly update the existing test to also test `ReplaceDict`.

**Comment by kshyatt** [comment]
Addressed comments and added some more methods for `GenericDict` in order to support `replace!` on it.

**Review by oscardssmith — APPROVED** [review]

File: test/sets.jl [hunk 4]
```diff
@@ -816,6 +816,22 @@ end
...
     @test_throws MethodError replace!(identity, 1=>2)
     @test_throws MethodError replace!(identity, 1=>2, 3=>4)
+
+    # test replace and friends for AbstractDicts
+    d1 = GenericDict(Dict(1=>2, 3=>4))
+    d2 = replace(d1, (1=>2) => (1=>"a"))
+    @test d2 == Dict(1=>"a", 3=>4)
+    @test d2 isa Dict{Int, Any}
+    replace!(d1, (1=>2) => (1=>-2))
+    @test d1 == GenericDict(Dict(1=>-2, 3=>4))
```
Comment (rfourquet) [inline test/sets.jl:826]: no need to wrap the rhs as a GenericDict
```suggestion
    @test d1 === replace!(d1, (1=>2) => (1=>-2))
    @test d1 == Dict(1=>-2, 3=>4)
```

File: test/sets.jl [hunk 5]
```diff
@@ -816,6 +816,22 @@ end
...
+
+    dd = Dict(1=>2, 3=>1, 5=>1, 7=>1)
+    for d1 in (dd, GenericDict(dd))
+        @test replace(d1, (1=>2) => (1=>"a"), count=0) == d1
+        d2 = replace((kv)->(kv[2] == 1 ? kv[1]=>2 : kv), d1, count=2)
```
Comment (rfourquet) [inline test/sets.jl:831]:
```suggestion
        d2 = replace(kv->(kv[2] == 1 ? kv[1]=>2 : kv), d1, count=2)
```

File: test/sets.jl [hunk 6]
```diff
@@ -816,6 +816,22 @@ end
...
+        @test count(v->v==2, values(d2)) == 3
+        @test count(v->v==1, values(d2)) == 1
```
Comment (rfourquet) [inline test/sets.jl:833]:
```suggestion
        @test count(==(2), values(d2)) == 3
        @test count(==(1), values(d2)) == 1
```

**Review by rfourquet — APPROVED** [review]
LGTM! (i just left minor style suggestions, feel free to ignore)

**Comment by kshyatt** [comment]
Test failures look unrelated although there sure are a lot of them...
